lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,754 @@
+# in the referenced literature but have been largely empirically determined and while a good
+# starting point could be could be further tuned.
+#
+# </div>
+
+# %%
+optimizer_init = {
+    "class_path": "torch.optim.AdamW",
+    "init_args": {"weight_decay": 1e-05, "eps": 1e-07, "lr": 1e-05},
+}
+
+# %% [markdown]
+# ### LR Scheduler Configuration
+#
+# <div id="a3">
+#
+# The [CosineAnnealingWarmRestarts scheduler](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html?highlight=cosineannealingwarm#torch.optim.lr_scheduler.CosineAnnealingWarmRestarts) nicely fits with our iterative finetuning since it does not depend upon a global max_epoch
+# value. The importance of initial warmup is reduced due to the innate warmup effect of Adam bias correction [[5]](#f3)
+# and the gradual thawing we are performing. Note that commonly used LR schedulers that depend on providing
+# max_iterations/epochs (e.g. the
+# [CosineWarmupScheduler](https://github.com/PyTorchLightning/lightning-tutorials/blob/0c325829101d5a6ebf32ed99bbf5b09badf04a59/course_UvA-DL/05-transformers-and-MH-attention/Transformers_MHAttention.py#L688)
+# used in other pytorch-lightning tutorials) also work with FinetuningScheduler. Though the LR scheduler is theoretically
+# justified [(Loshchilov & Hutter, 2016)](#f4), the particular values provided here are primarily empircally driven.
+#
+# </div>
+
+
+# %%
+lr_scheduler_init = {
+    "class_path": "torch.optim.lr_scheduler.CosineAnnealingWarmRestarts",
+    "init_args": {"T_0": 1, "T_mult": 2, "eta_min": 1e-07},
+}
+pl_lrs_cfg = {"interval": "epoch", "frequency": 1, "name": "CosineAnnealingWarmRestarts"}
+
+# %%
+# Load our lightning module...
+lightning_module_kwargs = {
+    "model_name_or_path": "microsoft/deberta-v3-base",
+    "optimizer_init": optimizer_init,
+    "lr_scheduler_init": lr_scheduler_init,
+    "pl_lrs_cfg": pl_lrs_cfg,
+}
+model = RteBoolqModule(**lightning_module_kwargs, experiment_tag="fts_explicit")
+
+# %% [markdown]
+# ### Callback Configuration
+#
+# The only callback required to invoke the ``FinetuningScheduler`` is the ``FinetuningScheduler`` callback itself.
+# Default versions of ``FTSCheckpoint`` and ``FTSEarlyStopping`` (if not specifying ``epoch_only_transitions``) will be
+# included ([as discussed above](#basic_usage)) if not provided in the callbacks list. For demonstration purposes I'm
+# including example configurations of all three callbacks below.
+
+# %%
+earlystopping_kwargs = {"monitor": "val_loss", "min_delta": 0.001, "patience": 2}
+checkpoint_kwargs = {"monitor": "val_loss", "save_top_k": 1}
+fts_kwargs = {"max_depth": 1}
+callbacks = [
+    FinetuningScheduler(ft_schedule=ft_schedule_name, **fts_kwargs),  # type: ignore # noqa
+    FTSEarlyStopping(**earlystopping_kwargs),  # type: ignore # noqa
+    FTSCheckpoint(**checkpoint_kwargs),  # type: ignore # noqa
+]
```

Comment:
```suggestion
callbacks = [
    FinetuningScheduler(ft_schedule=ft_schedule_name, max_depth=1),  # type: ignore # noqa
    FTSEarlyStopping(monitor="val_loss", min_delta=0.001, patience=2),  # type: ignore # noqa
    FTSCheckpoint(monitor="val_loss", save_top_k=1),  # type: ignore # noqa
]
```

Reply:
Better to remove the extra dict, this is a bit distracting. Better to rely on them when the arguments don't fit on a single line.

Reply:
In passing the same dict to all three scenarios, my intention was to highlight that all three scenarios were using these same config (plus I'm able to just pass these dicts in the two examples lower down rather than being more verbose). I def would agree if I was only passing the arguments once but that was my reasoning. Let me know if you still want me to explicitly pass all the arguments to each of the scenarios here and lower down though.